fix(ast/estree): make TS-only fields optional in TS type defs

overlookmotel opened the pull request:
When the source type is JS/JSX, `parseSync` returns an AST without any of the TS fields present (properties not present at all, not just defined as `undefined` / `null`). So to make our type defs match that AST (and also the TS AST which *does* have these fields), these fields need to be optional.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9846?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9846** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9846?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9846?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9844** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9844?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#9843** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9843?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9846).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@hi-ogawa I'm not a TypeScript user, and don't understand the TS syntax well. Can you please check this is correct?

The idea is that when the source type is JS/JSX, `parseSync` returns an AST without any of the TS fields present (properties not present at all, not just defined as `undefined` / `null`). So to make our type defs match that AST (and also the TS AST which *does* have these fields), I think these fields need to be optional.

hi-ogawa reviewed (approved):
Makes sense to me :+1: This can potentially give inconvenience for users expecting typescript estree types, which has some fields strictly defined (for examples `Class.implements` array https://github.com/typescript-eslint/typescript-eslint/blob/3efd99e954d1749225d65b774cfb3650a54ff083/packages/ast-spec/src/base/ClassBase.ts#L50), but this typing is more correct as oxc types.

Additionally, I think we can add jsdoc on each ts-only fields, so it's easier to tell which fields can be assumed to be strictly defined when users are exclusively using `astType: 'ts'` parser.

hi-ogawa commented:
### Merge activity

* **Mar 18, 3:39 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
fix(ast/estree): make TS-only fields optional in TS type defs (#9846)

When the source type is JS/JSX, `parseSync` returns an AST without any of the TS fields present (properties not present at all, not just defined as `undefined` / `null`). So to make our type defs match that AST (and also the TS AST which *does* have these fields), these fields need to be optional.

overlookmotel commented:
Thanks for reviewing. We could also codegen 2 sets of types for JS-only and with-TS ASTs. As we're codegen-ing everything, that wouldn't be too hard to do. But I don't know if it'd be an improvement or not.

hi-ogawa commented:
> Thanks for reviewing. We could also codegen 2 sets of types for JS-only and with-TS ASTs. As we're codegen-ing everything, that wouldn't be too much work.
> 
> But I don't know if it'd be an improvement or not, because `parseSync` returns one or the other depending on input. That's not possible to statically type, I don't think.

I don't think it's necessary right now, but just providing ts ast types can be useful without coupling with `parseSync` since that allows using that type mostly and only need `program as any as TsProgram` in one place.
